helm/chaos-mesh/templates/chaos-daemon-daemonset.yaml

```diff
@@ -80,10 +81,7 @@ spec:
             - /etc/chaos-daemon/cert/tls.key
           {{- end }}
           env:
-            {{- range $envKey, $envVal := .Values.chaosDaemon.env }}
-            - name: {{ $envKey | upper }}
-              value: {{ $envVal | quote }}
-            {{- end }}
+            {{- include "chaos-mesh.helpers.list-env-vars" .Values.chaosDaemon | indent 12 }}
```

Comment:
`indent xx` here will generate unnecessary whitespaces.

Reply:
Hmmm...weird behavior here. If the indent is removed, then the indentation of yaml will generate an error. But even if I write like this:

```yaml
          env:
{{- include "chaos-mesh.helpers.listEnvVars" .Values.dashboard | indent 12 }}
```

There will still be 12 spaces. It seems that this is a bug of helm. 🤔

Reply:
Well, it works fine on my local environment. (helm version: 3.5.3)

Could you show me some error messages?

Reply:
This helps me. I'm using helm `v3.8.1` in my local, it's updated by homebrew silently. I will try to downgrade helm and generate it again.

Reply:
Same behavior in `v3.5.2`. Remove `indent 12` will cause **Error: YAML parse error on chaos-mesh/templates/chaos-dashboard-deployment.yaml: error converting YAML to JSON: yaml: line 59: did not find expected key**

Then generated will be like:

```yaml
          env:
- name: CLEAN_SYNC_PERIOD
  value: "12h"
- name: DATABASE_DATASOURCE
  value: "/data/core.sqlite"
```

Reply:
hmmm, I wonder if you try this:

```yaml
          env:
            {{- include "chaos-mesh.helpers.listEnvVars" .Values.dashboard }}
```

or this:

```yaml
          env:
{{- include "chaos-mesh.helpers.listEnvVars" .Values.dashboard }}
```

Reply:
Oops, I found that I test the `chaos-daemon` on my local environment which the `env` is empty! 😿 

And I find a solution (got inspiration from [here](https://github.com/bitnami/charts/blob/c353b59c08c1e3026efff64d8d6d8561f0836670/bitnami/airflow/templates/_git_helpers.tpl)):
1. use `{{- end -}}` in the last line of a template
```yaml
{{- define "chaos-mesh.helpers.listEnvVars" -}}
...
{{- end -}}
```

2. use `trim` and `nindent` to control whitespaces:
```yaml
{{- include "chaos-mesh.helpers.listEnvVars" .Values.dashboard | trim | nindent 12 }}
```

BTW, I think keeping `{{- xxx -}}`(both `-`) in templates is a good practice, if we don't want to take more care of controlling whitespaces.

Reply:
Thx! It works. I will update this PR soon. 😄

Reply:
> BTW, I think keeping {{- xxx -}}(both -) in templates is a good practice, if we don't want to take more care of controlling whitespaces.

Makes sense. I just saw useless `-` at the end of some control statements and removed them.

Reply:
Resolved.

Reply:
A note: due to daemon has an value `env: {}` in `values.yaml`. The output will be:

```yaml
env:

  - name: TZ
     value: ${timezone}
```

Reply:
> A note: due to daemon has an value `env: {}` in `values.yaml`. The output will be:
> 
> ```yaml
> env:
> 
>   - name: TZ
>      value: ${timezone}
> ```

Maybe we could check whether the field is empty (like `{{ if .Values.dashboard.env }}`), and then decide to use the template.

Reply:
Completed. ✅ There should be no further questions now.